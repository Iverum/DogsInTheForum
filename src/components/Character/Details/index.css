input.error {
  background-color: #E74C3C;
}

aside {
  color: #969594;
}
