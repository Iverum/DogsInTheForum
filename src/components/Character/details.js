import React, { Component } from 'react'

export default class CharacterDetails extends Component {

  renderName(wrapperClass = 'row') {
    return (
      <div className={wrapperClass}>
        <label htmlFor="characterName">Name</label>
        <input className="u-full-width" type="text" placeholder="John Doe" id="characterName" />
      </div>
    )
  }

  renderBackground(wrapperClass = 'row') {
    return (
      <div className={wrapperClass}>
        <label htmlFor="characterBackground">Background</label>
        <input className="u-full-width" type="text" placeholder="Well Rounded" id="characterBackground" />
      </div>
    )
  }

  renderDescription(wrapperClass = 'six columns') {
    return (
      <div className={wrapperClass}>
        <label htmlFor="characterDescription">Description</label>
        <textarea className="u-full-width" placeholder="Very average" id="characterDescription"></textarea>
      </div>
    )
  }

  renderBasicInformation() {
    return (
      <div className="row">
        <div className="six columns">
          {this.renderName()}
          {this.renderBackground()}
        </div>
        {this.renderDescription()}
      </div>
    )
  }

  render() {
    return (
      <form className="twelve columns">
        {this.renderBasicInformation()}
        <input className="button-primary" type="submit" value="Update Character" />
      </form>
    )
  }

}
