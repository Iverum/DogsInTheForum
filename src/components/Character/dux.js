import _ from 'lodash'

// Actions
const ADD_CHARACTER = 'ditf/characters/ADD_CHARACTER'
const CLEAR_CHARACTERS = 'ditf/characters/CLEAR_CHARACTERS'

// Reducer
export default function reducer(state = [], action = {}) {
  switch (action.type) {
    case ADD_CHARACTER: {
      /*
       * This runs the risk of becoming slow as we add more characters. We
       * might want to consider indexing characters with a UUID in order to reduce the
       * time spent here.
       */
      if (_.find(state, character => character.uuid === action.data.uuid)) {
        return state
      }
      return [...state, action.data]
    }

    case CLEAR_CHARACTERS: {
      return []
    }

    default: {
      return state
    }
  }
}

// Action creators
export function addCharacter(character) {
  return {
    type: ADD_CHARACTER,
    data: character
  }
}

export function clear() {
  return { type: CLEAR_CHARACTERS }
}
