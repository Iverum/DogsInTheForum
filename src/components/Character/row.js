import React, { Component } from 'react'
import { Link } from 'react-router'

export default class CharacterRow extends Component {
  render() {
    return (
      <Link className='button u-full-width' to='characters/c4c1b0ad-4872-43cc-81b1-fb1a0ca77002'>
        {this.props.data.name}
      </Link>
    )
  }
}
