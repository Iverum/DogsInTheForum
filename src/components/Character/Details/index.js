import React, { Component } from 'react'
import firebase from 'firebase'
import { connect } from 'react-redux'
import './index.css'
import BasicInfo from './basic'
import Stats from './stats'
import Traits from './traits'
import Relationships from './relationships'
import Belongings from './belongings'
import * as characterActions from '../dux'

const emptyCharacter = {
  name: '',
  description: '',
  background: 'Well-Rounded',
  stats: {
    Acuity: { size: 'd6', number: 2 },
    Body: { size: 'd6', number: 2 },
    Heart: { size: 'd6', number: 2 },
    Will: { size: 'd6', number: 2 }
  },
  traits: [],
  relationships: [],
  belongings: []
}

const backgrounds = {
  'Well-Rounded': {
    'stats': { size: 'd6', number: 17 },
    'traits': [
      { size: 'd4', number: 1 },
      { size: 'd6', number: 4 },
      { size: 'd8', number: 2 }
    ],
    'relationships': [
      { size: 'd6', number: 4 },
      { size: 'd8', number: 2 }
    ]
  },
  'Strong History': {
    'stats': { size: 'd6', number: 13 },
    'traits': [
      { size: 'd6', number: 3 },
      { size: 'd8', number: 4 },
      { size: 'd10', number: 3 }
    ],
    'relationships': [
      { size: 'd4', number: 1 },
      { size: 'd6', number: 3 },
      { size: 'd8', number: 2 }
    ]
  },
  'Complicated History': {
    'stats': { size: 'd6', number: 15 },
    'traits': [
      { size: 'd4', number: 4 },
      { size: 'd6', number: 2 },
      { size: 'd10', number: 2 }
    ],
    'relationships': [
      { size: 'd6', number: 5 },
      { size: 'd8', number: 2 }
    ]
  },
  'Strong Community': {
    'stats': { size: 'd6', number: 13 },
    'traits': [
      { size: 'd4', number: 1 },
      { size: 'd6', number: 3 },
      { size: 'd8', number: 2 }
    ],
    'relationships': [
      { size: 'd6', number: 4 },
      { size: 'd8', number: 4 },
      { size: 'd10', number: 3 }
    ]
  },
  'Complicated Community': {
    'stats': { size: 'd6', number: 15 },
    'traits': [
      { size: 'd6', number: 6 },
      { size: 'd8', number: 2 }
    ],
    'relationships': [
      { size: 'd4', number: 4 },
      { size: 'd6', number: 2 },
      { size: 'd8', number: 2 },
      { size: 'd10', number: 2 }
    ]
  }
}

export class CharacterDetails extends Component {
  constructor(props) {
    super(props)
    this.updateCharacterFromDatabase = this.updateCharacterFromDatabase.bind(this)
    this.changeProperty = this.changeProperty.bind(this)
    this.state = { character: emptyCharacter }
  }

  componentWillMount() {
    const auth = firebase.auth()
    const database = firebase.database()
    auth.getRedirectResult()
      .then(() => {
        const id = auth.currentUser.uid
        this.characterRef = database.ref(`characters/${id}/${this.props.params.uuid}`)
        this.characterRef.off()
        this.characterRef.on('value', this.updateCharacterFromDatabase)
      })
  }

  componentWillUnmount() {
    this.characterRef.off()
    this.characterRef = null
  }

  updateCharacterFromDatabase(data) {
    const character = {
      ...emptyCharacter,
      ...data.val()
    }
    this.props.dispatch(characterActions.addCharacter(character))
    this.setState({ character })
  }

  changeProperty(property, value) {
    const newCharacter = {
      ...this.state.character,
      [property]: value
    }
    this.setState({
      character: newCharacter
    })
  }

  saveCharacter() {
    this.characterRef.set(this.state.character)
  }

  render() {
    const { background: characterBackground } = this.state.character
    return (
      <form className='twelve columns'>
        <BasicInfo
          {...this.state}
          onChange={this.changeProperty}
          backgrounds={backgrounds}
        />
        <hr />
        <div className='row'>
          <Stats
            {...this.state.character.stats}
            backgroundStats={backgrounds[characterBackground].stats}
            onChange={this.changeProperty}
          />
          <Traits
            attributes={this.state.character.traits}
            attributeDice={backgrounds[characterBackground].traits}
            onChange={this.changeProperty}
          />
        </div>
        <hr />
        <Relationships
          attributes={this.state.character.relationships}
          attributeDice={backgrounds[characterBackground].relationships}
          onChange={this.changeProperty}
        />
        <hr />
        <Belongings
          noDiceLimits
          attributes={this.state.character.belongings}
          onChange={this.changeProperty}
        />
        <hr />
        <input
          className='button-primary'
          type='submit'
          value='Update Character'
          onClick={this.saveCharacter.bind(this)}
        />
      </form>
    )
  }

}

export default connect()(CharacterDetails)
