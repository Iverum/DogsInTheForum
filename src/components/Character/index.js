import React, { Component } from 'react'
import { ListView, ListRows, Pagination } from 'react-list-combo'
import { connect } from 'react-redux'
import firebase from 'firebase'
import uuid from 'uuid/v4'
import CharacterRow from './row'
import * as characterActions from './dux'

export class Character extends Component {
  constructor(props) {
    super(props)
    this.createCharacter = this.createCharacter.bind(this)
    this.updateCharactersFromDatabase = this.updateCharactersFromDatabase.bind(this)
  }

  componentWillMount() {
    const auth = firebase.auth()
    const database = firebase.database()
    auth.getRedirectResult()
      .then(() => {
        const id = auth.currentUser.uid
        this.characterRef = database.ref(`character/${id}`)
        this.characterRef.off()
        this.characterRef.on('child_added', this.updateCharactersFromDatabase)
        this.characterRef.on('child_changed', this.updateCharactersFromDatabase)
      })
  }

  componentWillUnmount() {
    this.characterRef.off()
    this.characterRef = null
    this.props.dispatch(characterActions.clear())
  }

  updateCharactersFromDatabase(data) {
    const character = data.val()
    this.props.dispatch(characterActions.addCharacter(character))
  }

  createCharacter(name) {
    const newCharacter = {
      uuid: uuid(),
      name: 'New Character'
    }
    this.characterRef.child(`${newCharacter.uuid}`).set(newCharacter)
  }

  render() {
    console.log(this.props)
    return (
      <div className='twelve columns'>
        <header className='row'>
          <h1 className="nine columns">Characters</h1>
          <input
            onClick={this.createCharacter}
            className="button-primary three columns"
            type="button"
            value="Create New Character"
          />
        </header>
        <div className='row'>
          <ListView
            initData={this.props.characters}
            perPage={5}
          >
            <ListRows><CharacterRow /></ListRows>
            <Pagination className='pagination' />
          </ListView>
        </div>
      </div>
    )
  }

}

export default connect(state => ({
  characters: state.characters
}))(Character)
