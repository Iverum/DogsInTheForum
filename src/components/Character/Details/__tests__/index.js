import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import Character from '../index'

jest.mock('firebase', () => {
  return {
    auth: () => {
      return {
        currentUser: { uid: 'testUser' },
        getRedirectResult: () => { return Promise.resolve(true) }
      }
    },
    database: () => {
      return {
        ref: () => {
          return {
            off: () => {},
            on: () => {}
          }
        }
      }
    }
  }
})

test('character details are rendered correctly', () => {
  const component = renderer.create(
    <Character params={{ uuid: 'test' }} />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
