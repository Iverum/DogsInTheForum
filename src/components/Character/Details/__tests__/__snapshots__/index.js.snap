exports[`test character details are rendered correctly 1`] = `
<form
  className="twelve columns">
  <div
    className="row">
    <div
      className="six columns">
      <div
        className="row">
        <label
          htmlFor="characterName">
          Name
        </label>
        <input
          className="u-full-width"
          id="characterName"
          onChange={[Function]}
          placeholder="John Doe"
          type="text"
          value="" />
      </div>
      <div
        className="row">
        <label
          htmlFor="characterBackground">
          Background
        </label>
        <select
          className="u-full-width"
          id="characterBackground"
          onChange={[Function]}
          value="Well-Rounded">
          <option
            value="Well-Rounded">
            Well-Rounded
          </option>
          <option
            value="Strong History">
            Strong History
          </option>
          <option
            value="Complicated History">
            Complicated History
          </option>
          <option
            value="Strong Community">
            Strong Community
          </option>
          <option
            value="Complicated Community">
            Complicated Community
          </option>
        </select>
      </div>
    </div>
    <div
      className="six columns">
      <label
        htmlFor="characterDescription">
        Description
      </label>
      <textarea
        className="u-full-width"
        id="characterDescription"
        onChange={[Function]}
        placeholder="Very average"
        value="" />
    </div>
  </div>
  <hr />
  <div
    className="row">
    <div
      className="six columns">
      <div
        className="row">
        <h2
          className="three columns">
          Stats
        </h2>
        <aside
          className="nine columns">
          9d6 left to assign
        </aside>
      </div>
      <div
        className="row">
        <label
          className="two columns"
          htmlFor="statAcuity">
          Acuity
        </label>
        <input
          className="ten columns"
          id="statAcuity"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="2d6"
          type="text"
          value="2d6" />
      </div>
      <div
        className="row">
        <label
          className="two columns"
          htmlFor="statBody">
          Body
        </label>
        <input
          className="ten columns"
          id="statBody"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="2d6"
          type="text"
          value="2d6" />
      </div>
      <div
        className="row">
        <label
          className="two columns"
          htmlFor="statHeart">
          Heart
        </label>
        <input
          className="ten columns"
          id="statHeart"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="2d6"
          type="text"
          value="2d6" />
      </div>
      <div
        className="row">
        <label
          className="two columns"
          htmlFor="statWill">
          Will
        </label>
        <input
          className="ten columns"
          id="statWill"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="2d6"
          type="text"
          value="2d6" />
      </div>
    </div>
    <div
      className="six columns">
      <div
        className="row">
        <h2
          className="three columns">
          Traits
        </h2>
        <aside
          className="nine columns">
          1d4 4d6 2d8 left to assign
        </aside>
      </div>
      <input
        className="button"
        onClick={[Function]}
        type="button"
        value="New Trait" />
    </div>
  </div>
  <hr />
  <div
    className="row">
    <div
      className="row">
      <h2
        className="three columns">
        Relationships
      </h2>
      <aside
        className="nine columns">
        4d6 2d8 left to assign
      </aside>
    </div>
    <input
      className="button"
      onClick={[Function]}
      type="button"
      value="New Relationship" />
  </div>
  <hr />
  <div
    className="row">
    <div
      className="row">
      <h2
        className="three columns">
        Belongings
      </h2>
    </div>
    <input
      className="button"
      onClick={[Function]}
      type="button"
      value="New Belonging" />
  </div>
  <hr />
  <input
    className="button-primary"
    onClick={[Function]}
    type="submit"
    value="Update Character" />
</form>
`;
